api/src/main/java/org/apache/iceberg/types/Types.java

```diff
@@ -205,38 +208,74 @@ public String toString() {
   }
 
   public static class TimestampType extends PrimitiveType {
-    private static final TimestampType INSTANCE_WITH_ZONE = new TimestampType(true);
-    private static final TimestampType INSTANCE_WITHOUT_ZONE = new TimestampType(false);
 
+    private static final TimestampType INSTANCE_MICROS_WITH_ZONE =
+        new TimestampType(true, ChronoUnit.MICROS);
+    private static final TimestampType INSTANCE_MICROS_WITHOUT_ZONE =
+        new TimestampType(false, ChronoUnit.MICROS);
+    private static final TimestampType INSTANCE_NANOS_WITH_ZONE =
+        new TimestampType(true, ChronoUnit.NANOS);
+    private static final TimestampType INSTANCE_NANOS_WITHOUT_ZONE =
+        new TimestampType(false, ChronoUnit.NANOS);
+
+    /** @deprecated use {@link #microsWithZone()} instead. */
+    @Deprecated
     public static TimestampType withZone() {
-      return INSTANCE_WITH_ZONE;
+      return INSTANCE_MICROS_WITH_ZONE;
     }
 
+    /** @deprecated use {@link #microsWithoutZone()} instead. */
+    @Deprecated
     public static TimestampType withoutZone() {
-      return INSTANCE_WITHOUT_ZONE;
+      return INSTANCE_MICROS_WITHOUT_ZONE;
+    }
+
+    public static TimestampType microsWithZone() {
+      return INSTANCE_MICROS_WITH_ZONE;
+    }
+
+    public static TimestampType microsWithoutZone() {
+      return INSTANCE_MICROS_WITHOUT_ZONE;
+    }
+
+    public static TimestampType nanosWithZone() {
+      return INSTANCE_NANOS_WITH_ZONE;
+    }
+
+    public static TimestampType nanosWithoutZone() {
+      return INSTANCE_NANOS_WITHOUT_ZONE;
     }
 
     private final boolean adjustToUTC;
+    private final ChronoUnit unit;
 
-    private TimestampType(boolean adjustToUTC) {
+    private TimestampType(boolean adjustToUTC, ChronoUnit unit) {
       this.adjustToUTC = adjustToUTC;
+      this.unit = unit;
     }
 
     public boolean shouldAdjustToUTC() {
       return adjustToUTC;
     }
 
+    public ChronoUnit unit() {
```

Comment:
Why not use our own unit enum? Using ChronoUnit creates odd situations where you don't know in the code what unit you're going to get, even though there are only 2 possibilities. This also ties the Iceberg API to a Java API that may change and the cost of introducing our own enum is small.

Reply:
We use some ChronoUnit functionality, e.g. the `between` method. I can introduce our own enum wrapping a ChonoUnit and delegating that method, if you prefer.

Reply:
I pushed a draft commit 214ec93d6 switching to a custom `Unit` enum; I'll fold it into the main commit if you like it.

Reply:
I prefer not exposing ChronoUnit from the Iceberg API.

Reply:
Done.